sn_networking/src/relay_manager.rs

```diff
@@ -0,0 +1,172 @@
+    dialing_relays: BTreeSet<PeerId>,
+    candidates: BTreeMap<PeerId, Multiaddr>,
+}
+
+impl RelayManager {
+    pub(crate) fn new(initial_peers: Vec<Multiaddr>) -> Self {
+        let candidates = initial_peers
+            .into_iter()
+            .filter_map(|addr| {
+                for protocol in addr.iter() {
+                    if let Protocol::P2p(peer_id) = protocol {
+                        return Some((peer_id, addr));
+                    }
+                }
+                None
+            })
+            .collect();
+        Self {
+            connected_relays: Default::default(),
+            dialing_relays: Default::default(),
+            candidates,
+        }
+    }
+
+    pub(crate) fn add_potential_candidates(
+        &mut self,
+        peer_id: &PeerId,
+        addrs: &HashSet<Multiaddr>,
+        stream_protocols: &Vec<StreamProtocol>,
+    ) {
+        if self.candidates.len() >= MAX_POTENTIAL_CANDIDATES {
+            return;
+        }
+
+        if Self::does_it_support_relay_server_protocol(stream_protocols) {
+            // todo: collect and manage multiple addrs
+            if let Some(addr) = addrs.iter().next() {
+                // only consider non relayed peers
+                if !addr.iter().any(|p| p == Protocol::P2pCircuit) {
+                    self.candidates.insert(*peer_id, addr.clone());
+                }
+            }
+        }
+    }
+
+    /// Dials candidate relays.
+    pub(crate) fn dial_relays(&mut self, swarm: &mut Swarm<NodeBehaviour>) {
+        let mut n_dialed = self.connected_relays.len();
+
+        if n_dialed >= MAX_CONCURRENT_RELAY_CONNECTIONS {
+            return;
+        }
+
+        for (candidate_id, candidate_addr) in self.candidates.iter() {
+            match swarm.dial(
+                DialOpts::peer_id(candidate_id.clone())
+                    .condition(PeerCondition::NotDialing)
+                    // todo: should we add P2pCircuit here?
+                    // Just perform a direct connection and if the peer supports circuit protocol, the listen_on it.
+                    // the `listen_on(adrr.with P2pP)` will create the persistent relay connection right?
+                    .addresses(vec![candidate_addr.clone().with(Protocol::P2pCircuit)])
+                    .build(),
+            ) {
+                Ok(_) => {
+                    info!("Dialing Relay: {candidate_id:?} succeeded.");
+                    self.dialing_relays.insert(*candidate_id);
+                    n_dialed += 1;
+                    if n_dialed >= MAX_CONCURRENT_RELAY_CONNECTIONS {
+                        return;
+                    }
+                }
+                Err(err) => {
+                    error!("Error while dialing relay: {candidate_id:?} {err:?}",);
+                }
+            }
+        }
+    }
+
+    // todo: should we remove all our other `listen_addr`? Any should we block from adding `add_external_address` if
```

Comment:
hmmm we'd still want to be in `server` mode for libp2p though... So we may need to add that in 🤔 ?